[SOLVED] Compile error with ESP32

Hello Bert,
Long time no see. 
I was happy to see your MQTT library (V1.1.3). 
But when I try to compile a simple project for the ESP32 I get the following error: 

`
Compiling .pio\build\wemos_d1_mini32\libf63\espMqttClient\espMqttClient.cpp.o
.pio/libdeps/wemos_d1_mini32/espMqttClient/src/Transport/ClientSync.cpp: In member function 'virtual bool espMqttClientInternals::ClientSync::connect(IPAddress, uint16_t)':
.pio/libdeps/wemos_d1_mini32/espMqttClient/src/Transport/ClientSync.cpp:29:71: error: no matching function for call 
to 'WiFiClient::setSocketOption(int, int, int*, unsigned int)'
     client.setSocketOption(IPPROTO_TCP, TCP_NODELAY, &val, sizeof(int));
`
Do you have a hint for me?
Greetings, Siggi

--- a/examples/largepayload-esp8266/largepayload-esp8266.ino
+++ b/examples/largepayload-esp8266/largepayload-esp8266.ino
@@ -1,4 +1,6 @@
 #include <ESP8266WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClient.h>
 
 #define WIFI_SSID "yourSSID"
@@ -10,6 +12,8 @@
 WiFiEventHandler wifiConnectHandler;
 WiFiEventHandler wifiDisconnectHandler;
 espMqttClient mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 size_t fetchPayload(uint8_t* dest, size_t len, size_t index) {
   Serial.printf("filling buffer at index %zu\n", index);
@@ -33,7 +37,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void onWiFiConnect(const WiFiEventStationModeGotIP& event) {
@@ -56,7 +66,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -85,5 +96,10 @@ void setup() {
 }
 
 void loop() {
+  static currentMillis = millis();
+
   mqttClient.loop();
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
 }

--- a/examples/largepayload-esp8266/largepayload-esp8266.ino
+++ b/examples/largepayload-esp8266/largepayload-esp8266.ino
@@ -96,7 +96,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   mqttClient.loop();
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {

--- a/examples/ota-esp8266/ota-esp8266.ino
+++ b/examples/ota-esp8266/ota-esp8266.ino
@@ -142,7 +142,7 @@ void loop() {
     ESP.reset();
   }
 
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   mqttClient.loop();
 

--- a/examples/ota-esp8266/ota-esp8266.ino
+++ b/examples/ota-esp8266/ota-esp8266.ino
@@ -1,5 +1,7 @@
 #include <ESP8266WiFi.h>
 #include <Updater.h>
+#include <Ticker.h>
+
 #include <espMqttClient.h>
 
 #define WIFI_SSID "yourSSID"
@@ -13,6 +15,8 @@
 WiFiEventHandler wifiConnectHandler;
 WiFiEventHandler wifiDisconnectHandler;
 espMqttClient mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 bool disconnectFlag = false;
 bool restartFlag = false;
 
@@ -23,7 +27,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void onWiFiConnect(const WiFiEventStationModeGotIP& event) {
@@ -53,7 +63,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   }
   
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -125,16 +136,22 @@ void setup() {
 }
 
 void loop() {
-  mqttClient.loop();
-
-  if (disconnectFlag) {
-    // it's safe to call this multiple times
-    mqttClient.disconnect();
-  }
-
   if (restartFlag) {
     Serial.println("Rebooting... See you next time!");
     Serial.flush();
     ESP.reset();
+  }
+
+  static currentMillis = millis();
+
+  mqttClient.loop();
+
+  if (!disconnectFlag && reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
+
+  if (disconnectFlag) {
+    // it's safe to call this multiple times
+    mqttClient.disconnect();
   }
 }

--- a/examples/simple-esp32/simple-esp32.ino
+++ b/examples/simple-esp32/simple-esp32.ino
@@ -1,4 +1,6 @@
 #include <WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClient.h>
 
 #define WIFI_SSID "yourSSID"
@@ -8,6 +10,8 @@
 #define MQTT_PORT 1883
 
 espMqttClient mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 void connectToWiFi() {
   Serial.println("Connecting to Wi-Fi...");
@@ -16,7 +20,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void WiFiEvent(WiFiEvent_t event) {
@@ -57,7 +67,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -122,5 +133,9 @@ void setup() {
 }
 
 void loop() {
-  // nothing to do here
+  static currentMillis = millis();
+
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
 }

--- a/examples/simple-esp32/simple-esp32.ino
+++ b/examples/simple-esp32/simple-esp32.ino
@@ -133,7 +133,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
     connectToMqtt();

--- a/examples/simple-esp8266/simple-esp8266.ino
+++ b/examples/simple-esp8266/simple-esp8266.ino
@@ -128,7 +128,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   mqttClient.loop();
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {

--- a/examples/simple-esp8266/simple-esp8266.ino
+++ b/examples/simple-esp8266/simple-esp8266.ino
@@ -1,4 +1,6 @@
 #include <ESP8266WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClient.h>
 
 #define WIFI_SSID "yourSSID"
@@ -10,6 +12,8 @@
 WiFiEventHandler wifiConnectHandler;
 WiFiEventHandler wifiDisconnectHandler;
 espMqttClient mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 void connectToWiFi() {
   Serial.println("Connecting to Wi-Fi...");
@@ -18,7 +22,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void onWiFiConnect(const WiFiEventStationModeGotIP& event) {
@@ -51,7 +61,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -117,5 +128,10 @@ void setup() {
 }
 
 void loop() {
+  static currentMillis = millis();
+
   mqttClient.loop();
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
 }

--- a/examples/simpleAsync-esp32/simpleAsync-esp32.ino
+++ b/examples/simpleAsync-esp32/simpleAsync-esp32.ino
@@ -1,4 +1,6 @@
 #include <WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClientAsync.h>
 
 #define WIFI_SSID "yourSSID"
@@ -8,6 +10,8 @@
 #define MQTT_PORT 1883
 
 espMqttClientAsync mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 void connectToWiFi() {
   Serial.println("Connecting to Wi-Fi...");
@@ -16,7 +20,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void WiFiEvent(WiFiEvent_t event) {
@@ -57,7 +67,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -122,5 +133,9 @@ void setup() {
 }
 
 void loop() {
-  // nothing to do here
+  static currentMillis = millis();
+
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
 }

--- a/examples/simpleAsync-esp32/simpleAsync-esp32.ino
+++ b/examples/simpleAsync-esp32/simpleAsync-esp32.ino
@@ -133,7 +133,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
     connectToMqtt();

--- a/examples/simpleAsync-esp8266/simpleAsync-esp8266.ino
+++ b/examples/simpleAsync-esp8266/simpleAsync-esp8266.ino
@@ -1,4 +1,6 @@
 #include <ESP8266WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClientAsync.h>
 
 #define WIFI_SSID "yourSSID"
@@ -10,6 +12,8 @@
 WiFiEventHandler wifiConnectHandler;
 WiFiEventHandler wifiDisconnectHandler;
 espMqttClientAsync mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 void connectToWiFi() {
   Serial.println("Connecting to Wi-Fi...");
@@ -18,7 +22,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void onWiFiConnect(const WiFiEventStationModeGotIP& event) {
@@ -51,7 +61,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -117,5 +128,9 @@ void setup() {
 }
 
 void loop() {
-  // nothing to do here
+  static currentMillis = millis();
+
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
 }

--- a/examples/simpleAsync-esp8266/simpleAsync-esp8266.ino
+++ b/examples/simpleAsync-esp8266/simpleAsync-esp8266.ino
@@ -128,7 +128,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
     connectToMqtt();

--- a/examples/tls-esp32/tls-esp32.ino
+++ b/examples/tls-esp32/tls-esp32.ino
@@ -1,4 +1,6 @@
 #include <WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClient.h>
 
 #define WIFI_SSID "yourSSID"
@@ -15,6 +17,8 @@ const char rootCA[] = \
   "-----END CERTIFICATE-----\n";
 
 espMqttClientSecure mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 void connectToWiFi() {
   Serial.println("Connecting to Wi-Fi...");
@@ -23,7 +27,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void WiFiEvent(WiFiEvent_t event) {
@@ -61,7 +71,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -130,15 +141,22 @@ void setup() {
 }
 
 void loop() {
+  static currentMillis = millis();
+
+  mqttClient.loop();
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
+
   static uint32_t lastMillis = 0;
-  if (millis() - lastMillis > 5000) {
-    lastMillis = millis();
+  if (currentMillis - lastMillis > 5000) {
+    lastMillis = currentMillis;
     Serial.printf("heap: %u\n", ESP.getFreeHeap());
   }
 
   static uint32_t millisDisconnect = 0;
-  if (millis() - millisDisconnect > 60000) {
-    millisDisconnect = millis();
+  if (currentMillis - millisDisconnect > 60000) {
+    millisDisconnect = currentMillis;
     mqttClient.disconnect();
   }
 }

--- a/examples/tls-esp32/tls-esp32.ino
+++ b/examples/tls-esp32/tls-esp32.ino
@@ -141,7 +141,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   mqttClient.loop();
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {

--- a/examples/tls-esp32/tls-esp32.ino
+++ b/examples/tls-esp32/tls-esp32.ino
@@ -143,7 +143,6 @@ void setup() {
 void loop() {
   static uint32_t currentMillis = millis();
 
-  mqttClient.loop();
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
     connectToMqtt();
   }

--- a/examples/tls-esp8266/tls-esp8266.ino
+++ b/examples/tls-esp8266/tls-esp8266.ino
@@ -132,7 +132,7 @@ void setup() {
 }
 
 void loop() {
-  static currentMillis = millis();
+  static uint32_t currentMillis = millis();
 
   mqttClient.loop();
   if (reconnectMqtt && currentMillis - lastReconnect > 5000) {

--- a/examples/tls-esp8266/tls-esp8266.ino
+++ b/examples/tls-esp8266/tls-esp8266.ino
@@ -1,4 +1,6 @@
 #include <ESP8266WiFi.h>
+#include <Ticker.h>
+
 #include <espMqttClient.h>
 
 #define WIFI_SSID "yourSSID"
@@ -13,6 +15,8 @@ const uint8_t fingerprint[] = {0xee, 0xbc, 0x4b, 0xf8, 0x57, 0xe3, 0xd3, 0xe4, 0
 WiFiEventHandler wifiConnectHandler;
 WiFiEventHandler wifiDisconnectHandler;
 espMqttClientSecure mqttClient;
+bool reconnectMqtt = false;
+uint32_t lastReconnect = 0;
 
 void connectToWiFi() {
   Serial.println("Connecting to Wi-Fi...");
@@ -21,7 +25,13 @@ void connectToWiFi() {
 
 void connectToMqtt() {
   Serial.println("Connecting to MQTT...");
-  mqttClient.connect();
+  if (!mqttClient.connect()) {
+    reconnectMqtt = true;
+    lastReconnect = millis();
+    Serial.println("Connecting failed.");
+  } else {
+    reconnectMqtt = false;
+  }
 }
 
 void onWiFiConnect(const WiFiEventStationModeGotIP& event) {
@@ -54,7 +64,8 @@ void onMqttDisconnect(espMqttClientTypes::DisconnectReason reason) {
   Serial.printf("Disconnected from MQTT: %u.\n", static_cast<uint8_t>(reason));
 
   if (WiFi.isConnected()) {
-    connectToMqtt();
+    reconnectMqtt = true;
+    lastReconnect = millis();
   }
 }
 
@@ -121,5 +132,10 @@ void setup() {
 }
 
 void loop() {
+  static currentMillis = millis();
+
   mqttClient.loop();
+  if (reconnectMqtt && currentMillis - lastReconnect > 5000) {
+    connectToMqtt();
+  }
 }

